Implement scroll bar

Tyriar opened the pull request:
Fixes #77
Fixes #149

### Implementation

The scroll bar is implemented using a `.xterm-scroll-area` element which is resized based on what the entire buffer is, this is contained within `.xterm-viewport` which uses `overflow-y: scroll` to show a native scroll bar. The viewport is kept separate from the rows in order to prevent calls to `refresh` triggering `scroll` events on the viewport.

Scroll and wheel events are then delegated to the viewport which in turn passes it on to the terminal. This can happen in two ways, events originating from the viewport (mouse wheel, drag and drop scroll bar, up/down buttons on scroll bar) and events originating from the terminal (scrolling via keyboard, new output, etc.).

The implementation supports various scroll configurations, properly utilizing `WheelEvent.deltaY` based on the [`WheelEvent.deltaMode`](https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent/deltaMode), this allows certain configurations to specify whether the user prefers to scroll by pixels, lines or pages.

### Remaining tasks

- [x] Write some tests
- [x] Manually test on Windows
  - [x] Edge
  - [x] Chrome
  - [x] Firefox
  - [x] IE11
- [ ] See if anything can be improved with OS X scroll bar visibility
- [ ] Allow custom `line-height`?
- [x] Disable/fill scroll bar when in an application such as `vim`

### Notes

- I dropped support for `mousewheel` and `DOMMouseScroll` events in favor of `WheelEvent` since it's just complicating the support and xterm.js' [supported browsers](https://github.com/sourcelair/xterm.js#browser-support) are well within [the browsers supporting `WheelEvent`](https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent#Browser_compatibility)
- Eventually a bunch of `Terminal`'s variables (`ydisp`, `y`, `ybase`) could be pulled into `Viewport` to simplify code, the virtual selection implementation would probably best live here too.
- Safari's scroll bar seems a little glitchy around drag of the scroll bar. This is not a big issue though considering it's probably not a very common use case on Macs.

### What it looks like

Ubuntu 16.04/Chrome 52:

![image](https://cloud.githubusercontent.com/assets/2193314/17381587/7f6a24de-5981-11e6-8820-902aaf52106d.png)

Ubuntu 16.04/Firefox 47

![image](https://cloud.githubusercontent.com/assets/2193314/17382044/dea4e09a-5983-11e6-862d-089311a5a9c0.png)

Windows 10/Edge 14

![image](https://cloud.githubusercontent.com/assets/2193314/17424710/81d7d8ba-5a7b-11e6-989d-d390c26dc058.png)

parisk commented:
The scrollbar works great in OS X Chrome. A couple of things I noticed that it would be great if they were fixed:

- Clicking on the scrollbar focuses on the terminal's parent element, which gets an outline styling (see attached screenshot)
- The scrollbar appears only after scrolling first using the trackpad (like before), while on OS X if a scrollbar exists, it should appear even when hovering the corresponding element

## Screenshot
![screenshot](https://photos-3.dropbox.com/t/2/AADytPJ6GEjQeqhizAa_8uPX-cYL9kHRt9tnGSe2adP6GA/12/33160981/png/32x32/3/1470322800/0/2/Fullscreen_04_08_16_10_57.png/ENrukRkYs-wBIAcoBw/Y-XbAPEQEV8eQnMFUNQKQGJdIAoZnQCHG3aAeMsxfxc?size_mode=3&size=1600x1200)

parisk commented:
Thanks a lot for this @Tyriar 🎉 .

soichih commented:
How can I try this? Cloning the master doesn't seem to add the scrollbar. Is there any instruction?

parisk commented:
Hi @soichih! You have to checkout this branch at @Tyriar's fork. To do this, run the following commands in your terminal:

```
git checkout -b Tyriar-77_scroll_bar master
git pull https://github.com/Tyriar/xterm.js.git 77_scroll_bar
```

Let us know if you ran into any trouble.

Tyriar commented:
@parisk RE your comments:

- I added `.terminal:focus` rule to remove outline
- It seems to be the platform convention that hovering over the side does not trigger the scrollbar, tested on Chrome, Outlook and Safari.

I fixed the issue with the transparent scroll bar on OS X. This is ready to merge imo, let me know what you think.

parisk commented:
👍  awesome. LGTM.

Feel free to merge!
